File: frontend/src/types/worker.ts
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
// Types
import { AttributeT } from './attribute';

dayjs.extend(utc);

// ─── Weekly Preferences ──────────────────────────────────────────────────────

export type WeekParity = 'all' | 'even' | 'odd';

export type SlotRestriction = 'no_work' | 'no_normal' | 'no_duty' | 'no_specific';

export type WeeklySlotPreference = {
  dayOfWeek: number; // 0=Monday … 6=Sunday
  slot: 'morning' | 'afternoon' | 'night';
  restriction: SlotRestriction;
  shiftIds: string[];
  weekParity: WeekParity;
};

export type WeeklyPreferences = {
  enabled: boolean;
  slots: WeeklySlotPreference[];
};

// ─── Worker ───────────────────────────────────────────────────────────────────

export type WorkerT = {
  id: string;
  teamId: string;
  name: string;
  acronym: string;
  acronymCustom: boolean;
  employmentStartDate: dayjs.Dayjs;
  employmentEndDate: dayjs.Dayjs | null;
  weeklyHours: number;
  weeklyHoursDesired: number;
  dutiesPerMonth: number;
  annualLeave: number;
  specialtyIds: string[];
  deleted: boolean;
  userId: string | null;
  attributes: AttributeT[];
  weeklyPreferences?: WeeklyPreferences;
};

/**
 * Convert raw API data to WorkerT
 */
export function toWorkerT(data: any): WorkerT {
  return {
    ...data,
    employmentStartDate: dayjs.unix(data.employmentStartDate).utc(),
    employmentEndDate: data.employmentEndDate ? dayjs.unix(data.employmentEndDate).utc() : null,
  };
}

/**
 * Convert WorkerT to API request payload
 */
export function fromWorkerT(data: WorkerT): any {
  return {
    ...data,
    employmentStartDate: data.employmentStartDate.unix(),
    employmentEndDate: data.employmentEndDate ? data.employmentEndDate.unix() : null,
  };
}

File: backend/shared/src/shared/schemas/dto/worker.py
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field

from shared.schemas.dto.attribute import AttributeDTO


class WeeklySlotPreferenceDTO(BaseModel):
    dayOfWeek: int
    slot: str
    restriction: str
    shiftIds: List[str] = []
    weekParity: str = "all"


class WeeklyPreferencesDTO(BaseModel):
    enabled: bool = False
    slots: List[WeeklySlotPreferenceDTO] = []


class WorkerDTO(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": "worker-1",
                "teamId": "64e9b7f1e13e4a1a9c8b4567",
                "name": "Jane Smith",
                "acronym": "JS",
                "acronymCustom": False,
                "employmentStartDate": 1700000000.0,
                "employmentEndDate": None,
                "weeklyHours": 40,
                "weeklyHoursDesired": 36,
                "dutiesPerMonth": 10,
                "annualLeave": 25,
                "deleted": False,
                "specialtyIds": ["specialty-1"],
                "userId": None,
                "attributes": [],
                "weeklyPreferences": None,
            }
        }
    )

    id: str
    teamId: str
    name: str
    acronym: str
    acronymCustom: bool = Field(examples=[False])
    employmentStartDate: float = Field(examples=[1700000000.0])
    employmentEndDate: float | None
    weeklyHours: int = Field(examples=[40])
    weeklyHoursDesired: int = Field(examples=[36])
    dutiesPerMonth: int = Field(examples=[10])
    annualLeave: int = Field(examples=[25])
    deleted: bool = Field(examples=[False])
    specialtyIds: List[str] = Field(examples=[["specialty-1"]])
    userId: str | None
    attributes: List[AttributeDTO]
    weeklyPreferences: Optional[WeeklyPreferencesDTO] = None
